File: Python3.x/17-Letter_Combinations_of_a_Phone_Number.py
# 递归算法
class Solution:
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if len(digits) == 0: return []
        phone = {2: 'abc', 3: 'def', 4: 'ghi', 5: 'jkl', 6: 'mno', 7: 'pqrs', 8: 'tuv', 9: 'wxyz'}
        tmp_list = ['']
        for i in map(int, list(digits)):
            tmp = []
            for j in tmp_list:
                for k in phone[i]:
                    tmp.append(j + k)
            tmp_list = tmp
        return tmp_list

File: Algorithms/C++/17-Letter_Combinations_of_a_Phone_Number.cpp
// Runtime: 4 ms, faster than 59.36% of C++ online submissions for Letter Combinations of a Phone Number.
// Memory Usage: 8.7 MB, less than 71.43% of C++ online submissions for Letter Combinations of a Phone Number.
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        return recursion(digits, res);
    }
private:
    vector<string> phone = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> recursion(string digits, vector<string> cur) {
        if (digits.size() == 0)
            return cur;
        vector<string> next;
        int number = digits[0] - '0';
        if (cur.size() == 0)
            for (char c : phone[number])
                next.push_back(string(1, c));
        else
            for (char c : phone[number])
                for (string s : cur)
                    next.push_back(s + c);
        return recursion(digits.substr(1, digits.size()), next);
    }
};
